server/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -442,6 +444,9 @@ private static ElasticsearchException innerFromXContent(XContentParser parser, b
                     reason = parser.text();
                 } else if (STACK_TRACE.equals(currentFieldName)) {
                     stack = parser.text();
+                } else if (EsRejectedExecutionException.EXECUTOR_SHUTDOWN.equals(currentFieldName)) {
```

Comment:
instead of parsing back this exception into its own type, which we don't do for a lot of other exceptions, why don't we adapt the `BulkProcessor` to retry based on the status returned with the exception? That should work for both transport client and REST client I think. I think that configuring the exception to be retried on is not necessary, as we only ever use it for rejections (or we could make the status configurable instead of the class), we should check if the root cause is an `ElasticsearchException` , if so cast and check the returned `status`.

This needs testing, for instance porting the existing `BulkProcessorRetryIT` to the rest-high-level tests and adapting it similar to what I am doing in #29263 for `BulkProcessorIT`.

One other thing that I noticed is that when `canRetry` returns  `true`, we will retry all failed items from that response, but we don't check again the status code, nor the exception type. As a follow-up, we may want to fix that in `createBulkRequestForRetry`.